core/runtime.rs

```diff
@@ -349,6 +358,41 @@ impl JsRuntime {
       .unwrap();
   }
 
+  /// Initializes JS of provided Extensions
+  // NOTE: this will probably change when streamlining snapshot flow
+  pub fn init_extension_js(&mut self) -> Result<(), AnyError> {
+    // Take extensions to avoid double-borrow
+    let mut extensions: Vec<Extension> = self.extensions.drain(..).collect();
+    for m in extensions.iter_mut() {
+      let js_files = m.init_js();
+      for (filename, source) in js_files {
+        self.execute_static(filename, source)?;
+      }
+    }
+    self.extensions = extensions;
+    Ok(())
+  }
+
+  /// Initializes ops of provided Extensions
+  // NOTE: this will probably change when streamlining snapshot flow
+  pub fn init_extension_ops(&mut self) -> Result<(), AnyError> {
+    let op_state = self.op_state();
+    // Original OpRegistrar
+    let mut registrar: RcOpRegistrar =
+      Rc::new(RefCell::new(JsRuntimeOpRegistrar(op_state.clone())));
+
+    // Wrap registrar
+    for e in self.extensions.iter_mut() {
+      registrar = e.init_registrar(registrar.clone());
+    }
```

Comment:
I'm confused about the flow of registrars and middleware.

Most of the Extensions do not register any middleware - EG op_crates/url - however somehow they all have Metrics applied to them? From what I can tell the init function in runtime/metrics.rs creates an Extension with middleware, and it is loaded as the last Extension for Worker/WebWorker. Does order matter in registering of extensions? It seems to be the case here.

Reply:
@ry Order of extensions does NOT matter in regards to middleware, it basically inits all the middleware provided by all extensions before registering any of the ops.

Reply:
I'm also confused what's going on here. Why does `init_registrar` returns another registrar? It seems that it happens for every registered extension. What's the use case?

Reply:
It's essentially a reduce, transforming our initial OpRegistrar into a fully wrapped one.

We then register all ops using that fully wrapped OpRegistrar, so middleware provided by any of the extensions gets applied to all ops.

Reply:
This seems like it could get a little hairy if you have more than a few middlewares. Each middleware is going to Box and RefCell its parent registrar (which is also Boxed and RefCelled). It's effectively a recursive queue of `Rc<RefCell<OpRegistrar>>`. Lots of needless pointer chasing and `borrow_mut`, as well as using a bunch of stack space for the recursive function calls.

Probably worth checking, but I'd guess that collecting the `Box<OpMiddlewareFn>` into a `Vec` and then making an `OpRegistrar` which iterates over and executes these `OpMiddlewareFn` before calling into the parent `OpRegistrar` will be faster and less allocation heavy. I also think it's easier to reason about but that's possibly just me :shrug: .

Something like this, which probably doesn't compile but hopefully you get the idea

```rust
impl Extension {
    // ...
    pub fn middleware_fn(&mut self) -> Option<Box<OpMiddlewareFn>> { /* ... */ }
    // ...
}

struct JsRuntimeOpRegistrar<'a> {
    middlewares: Vec<Box<OpMiddlewareFn>>,
    state: &'a mut OpState
}
impl<'a> OpRegistrar for JsRuntimeOpRegistrar<'a> {
    fn register_op(&mut self, name: &str, op_fn: Box<OpFn>) -> OpId {
        let mut new_op = op_fn;
        for middleware in &self.middlewares {
            new_op = middleware(name, new_op);
        }
        self.state.op_table.register_op(name, new_op);
    }
}

pub fn init_extension_ops(&mut self) -> Result<(), AnyError> {
    let op_state = self.op_state();
    let mut op_state = op_state.borrow_mut();
    let mut registrar = JsRuntimeOpRegistrar {
        middlewares: vec![],
        state: &mut *op_state,
    };
    for e in self.extensions.iter_mut() {
        if let Some(op) = e.middleware_fn() {
            registrar.middlewares.push(op);
        }
    }
    let registrar = Rc::new(RefCell::new(registrar));
    // ... continue as before
}
```

The same criticism can probably be applied to the `OpMiddlewareFn` effectively doing the same thing as well, I suppose, but that would probably require a much larger rewrite, from what I can see.